src/server/ua_server_async.c

```diff
@@ -250,19 +251,11 @@ UA_AsyncManager_createAsyncOp(UA_AsyncManager *am, UA_Server *server,
     }
 
     UA_AsyncOperation *ao = (UA_AsyncOperation*)UA_calloc(1, sizeof(UA_AsyncOperation));
-    if(!ao) {
-        UA_LOG_ERROR(&server->config.logger, UA_LOGCATEGORY_SERVER,
-                     "UA_Server_SetNextAsyncMethod: Mem alloc failed.");
-        return UA_STATUSCODE_BADOUTOFMEMORY;
-    }
+    UA_CHECK_MEM(ao, goto mem_error);
```

Comment:
Now it gets interesting.
Here we could have to specialize UA_CHECK_MEM_ERROR to remove the goto.
Or we de-specialize UA_CHECK_ERROR to include a logical statement instead of just the variable name.

Reply:
Yes, I also thought about that, there are some ideas:

1. Have a specific base macro that just asserts that the given boolean expression is not true.
```c
#define UA_CHECK(A, EVAL) if(!(A)) { EVAL; }

// usage:
void *d = malloc(...);
UA_CHECK(d, return error);

UA_StatusCode rv = do_some();
UA_CHECK(rv == UA_STATUSCODE_GOOD, return error);
```
This would be good for general use of the check macro in places where the status code checking is not that extensively used.

2. Have a specific macro to check the status (like you suggested earlier), referencing the base asserter.

```c
#define UA_CHECK_STATUS(STATUS, EVAL) UA_CHECK((STATUS) == UA_STATUSCODE_GOOD, EVAL)
```

3. Have specific macros for exactly those two cases that include logging for ERROR, INFO, WARNING. this would mean 6 macros for the bare assertion and the one including the error code check. Downside: Macronames kind of explode.

```c
#define UA_CHECK_STATUS_ERROR(...)
#define UA_CHECK_ERROR(...)

#define UA_CHECK_STATUS_INFO(...)
#define UA_CHECK_INFO(...)

#define UA_CHECK_STATUS_WARN(...)
#define UA_CHECK_WARN(...)
```

Or we leave out all the specifications for statuscode and just use the explicit versions. The downside is, that there are about a 1000 uses of `if (... != UA_STATUSCODE_GOOD)` in the code until now so it could reduce at least some of that verbosity.

Reply:
Although it might be no such big change in verbosity, but it would at least encapslate the specific error checking. E.g. if indeed one might want to check the statuscode in an augmented way than "== UA_STATUSCODE_GOOD".

```c
UA_CHECK_ERROR(rv == UA_STATUSCODE_GOOD, return rv, logger, CATEGORY, "message");

// versus

UA_CHECK_STATUS_ERROR(rv, return rv, logger, CATEGORY, "message");
```

Reply:
A huge pro for specific STATUS check macros:
  `UA_StatusCode_name(...)` can be added to the log information without any further effort.

Reply:
> UA_CHECK_STATUS_ERROR(rv, return rv, logger, CATEGORY, "message");

Right, UA_CHECK_STATUS_ERROR is a good macro to have for the StatusCode name.

It seems that each macro only takes line or so to define.
With the macro nesting you already have in place.

So I would say let's add **all of the above**. ;-)